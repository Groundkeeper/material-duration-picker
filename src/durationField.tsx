import {TextField, TextFieldProps} from "@material-ui/core";
import * as React from 'react'
import {ComponentType, useEffect, useState} from "react";

export type DurationFieldProps = TextFieldProps & {
  value: number | null
  onConfirm: (value: number | null) => void;
  TextFieldComp?: ComponentType<TextFieldProps>;
}

export const DurationField = ({value: _value, onConfirm, TextFieldComp = TextField, ...props}: DurationFieldProps) => {
  const [value, setValue] = useState(_value)
  useEffect(() => {
    setValue(_value)
  }, [_value, setValue])

  return (
    <TextFieldComp
      type="number"
      onChange={({target}) => {
        const num = +target.value
        setValue(isNaN(num) || !target.value ? null : num);
      }}
      onBlur={() => {
        onConfirm(value)
      }}
      value={value ?? ''}
      fullWidth
      {...props}
    />
  )
}
