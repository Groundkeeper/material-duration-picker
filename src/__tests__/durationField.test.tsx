import {fireEvent, render, screen, waitFor} from "@testing-library/react";
import * as React from "react";
import {DurationField} from "../durationField";

describe('durationField', ()=>{

  it('can match snapshot', async () => {
    await waitFor(() => {
      expect(
        render(
          <DurationField
            view={"minutes"}
            duration={{minutes: 30}}
            onConfirm={() => {
            }}
          />
        ).asFragment()
      ).toMatchSnapshot();
    });
  })

  it('can match snapshot with custom labels', async () => {
    await waitFor(() => {
      expect(
        render(
          <DurationField
            labels={{
              minutes: 'CUSTOM_LABEL'
            }}
            view={"minutes"}
            duration={{minutes: 30}}
            onConfirm={() => {
            }}
          />
        ).asFragment()
      ).toMatchSnapshot();
    });
  })


  it('can confirm value when value is invalid', async () => {

    const mockOnConfirm = jest.fn()

    const utils = render(
      <DurationField
        view={"minutes"}
        duration={{minutes: 30}}
        onConfirm={mockOnConfirm}
        inputProps={{
          title: 'fieldTitle'
        }}
        />
    )

    const field: any= utils.getByTitle('fieldTitle')

    expect(field.value).toBe('30')
    fireEvent.change(field, {target: {value: NaN}})
    fireEvent.blur(field)
    expect(field.value).toBe('')
    expect(mockOnConfirm).toHaveBeenCalledWith(undefined)
  })
})
