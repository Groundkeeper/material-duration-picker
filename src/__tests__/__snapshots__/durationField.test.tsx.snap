// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`durationField can match snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
      data-shrink="true"
    >
      Minutes
    </label>
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiInput-input"
        type="number"
        value="30"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`durationField can match snapshot with custom labels 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
      data-shrink="true"
    >
      CUSTOM_LABEL
    </label>
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiInput-input"
        type="number"
        value="30"
      />
    </div>
  </div>
</DocumentFragment>
`;
